codex/sales/salesagent.nim

```diff
@@ -0,0 +1,85 @@
+import pkg/chronos
+import pkg/upraises
+import pkg/stint
+import ./statemachine
+import ./states/[downloading, unknown]
+import ../contracts/requests
+import ../rng
+
+proc newSalesAgent*(sales: Sales,
+                    requestId: RequestId,
+                    availability: ?Availability,
+                    request: ?StorageRequest): SalesAgent =
+  SalesAgent(
+    sales: sales,
+    requestId: requestId,
+    availability: availability,
+    request: request)
+
+# fwd declarations
+proc subscribeCancellation*(agent: SalesAgent): Future[void] {.gcsafe.}
+proc subscribeFailure*(agent: SalesAgent): Future[void] {.gcsafe.}
+
+proc populateRequest(agent: SalesAgent) {.async.} =
+  if agent.request.isNone:
+    agent.request = await agent.sales.market.getRequest(agent.requestId)
+
+proc init*(agent: SalesAgent, numSlots: uint64) {.async.} =
```

Comment:
Can we name this `start` (and rename `deinit` to `stop`)?

Reply:
Done